### ISSUE TITLE
Privacy consideration: prerendered origin gains access to initial origin's heuristics

### ISSUE THREAD

**Issue by noamr** [WICG/nav-speculation#106]
Given that a URL at origin A prerenders a URL at origin B:

The privacy considerations in the spec only mention that information from origin (B) may leak to origin (A).
However, when speculative prerendering is considered, private information from origin (A) can leak to origin (A).

Assuming that URL for (B) has some user identifiers (such as `fbclid`), a triggered prerender of (B) can divulge information about the user's behavior in origin (A).
As a rudimentary example: If the heuristics of prerendering are to only prerender the URL when the corresponding link is visible, and the link is only visible when the user is logged in, then a prerender request to origin (B) would be a signal to origin (B) that the user is logged in to origin (A) - without explicit action from origin (A) that sends that information.

The owner of origin (A) could mitigate this with `speculationrules`, however since a lot of the heuristics are internal to the UA, it's not always clear how to do this safely, and this becomes a bit of a cumbersome opt-out.

**Comment by jeremyroman** [comment]
This is a good point. Some mitigating factors:
* cross-origin fetches wouldn't have credentials (so this is less identifying than it would otherwise be)
* user agents can treat cross-origin URLs differently to reduce the extent to which this is useful
* because prefetches and prerenders have a cost, UAs are likely to not do an unbounded number of them

**Comment by noamr** [comment]
> This is a good point. Some mitigating factors:
> 
> * cross-origin fetches wouldn't have credentials (so this is less identifying than it would otherwise be)
> * because prefetches and prerenders have a cost, UAs are likely to not do an unbounded number of them

These steps make the problem slightly smaller but still leave a privacy breach surface.

> * user agents can treat cross-origin URLs differently to reduce the extent to which this is useful

Yes, and I think this different treatment should be explicitly specified. I don't have a clear vision of how this would work. Maybe  allowing only imperative APIs for cross-origin or making the rules more explicit.

I think this may affect the shape of `speculationrules`. Even though the first step is to support same-origin (which doesn't have this problem), we don't want to get to a point where we have to change the API shape due to this potential privacy leak.

**Comment by jeremyroman** [comment]
> Even though the first step is to support same-origin (which doesn't have this problem)

On the prerendering side, this is true. For prefetching, this isn't the case -- cross-origin prefetch doesn't have as many complications as cross-origin prerendering.

**Comment by noamr** [comment]
> > Even though the first step is to support same-origin (which doesn't have this problem)
> 
> On the prerendering side, this is true. For prefetching, this isn't the case -- cross-origin prefetch doesn't have as many complications as cross-origin prerendering.

Right, to be clear - this issue only refers to prerendering and not to prefetching.
